src/openrct2/world/sprite.c

```diff
@@ -823,3 +830,107 @@ bool sprite_get_flashing(rct_sprite *sprite)
     assert(sprite->unknown.sprite_index < MAX_SPRITES);
     return _spriteFlashingList[sprite->unknown.sprite_index];
 }
+
+static bool sprite_is_in_cycle(uint16 sprite_idx)
+{
+    if (sprite_idx == SPRITE_INDEX_NULL)
+    {
+        return false;
+    }
+    const rct_sprite * fast = get_sprite(sprite_idx);
+    const rct_sprite * slow = fast;
+    bool increment_slow = false;
+    bool cycled = false;
+    while (fast->unknown.sprite_index != SPRITE_INDEX_NULL)
+    {
+        // increment fast every time, unless reached the end
+        if (fast->unknown.next == SPRITE_INDEX_NULL)
+        {
+            break;
+        }
+        else {
+            fast = get_sprite(fast->unknown.next);
+        }
+        // increment slow only every second iteration
+        if (increment_slow)
+        {
+            slow = get_sprite(slow->unknown.next);
+        }
+        increment_slow = !increment_slow;
+        if (fast == slow)
+        {
+            cycled = true;
+            break;
+        }
+    }
+    return cycled;
+}
+
+static bool sprite_is_in_quadrant_cycle(uint16 sprite_idx)
+{
+    if (sprite_idx == SPRITE_INDEX_NULL)
+    {
+        return false;
+    }
+    const rct_sprite * fast = get_sprite(sprite_idx);
+    const rct_sprite * slow = fast;
+    bool increment_slow = false;
+    bool cycled = false;
+    while (fast->unknown.sprite_index != SPRITE_INDEX_NULL)
+    {
+        // increment fast every time, unless reached the end
+        if (fast->unknown.next_in_quadrant == SPRITE_INDEX_NULL)
+        {
+            break;
+        }
+        else {
+            fast = get_sprite(fast->unknown.next_in_quadrant);
+        }
+        // increment slow only every second iteration
+        if (increment_slow)
+        {
+            slow = get_sprite(slow->unknown.next_in_quadrant);
+        }
+        increment_slow = !increment_slow;
+        if (fast == slow)
+        {
+            cycled = true;
+            break;
+        }
+    }
+    return cycled;
+}
+
+bool check_for_sprite_list_cycles(bool fix)
+{
+    for (sint32 i = 0; i < NUM_SPRITE_LISTS; i++) {
+        if (sprite_is_in_cycle(gSpriteListHead[i])) {
+            if (fix)
+            {
+                rct_sprite * spr = get_sprite(gSpriteListHead[i]);
+
+                // There may be a better solution than simply setting this to 0xFFFF
+                spr->unknown.next = SPRITE_INDEX_NULL;
```

Comment:
Thinking about it a bit more, this is not a proper fix.

Imagine such a list:
```
A--->B--->C--->D--->E
     ^              |
     |              |
     +--------------+
```

You will only have removed node `A` from it, while there is still a loop present. You need to fix `slow` (or `fast`, but it is the same node at the time it detects the cycle), not the head.